text/0091-ci-upload-tokens.md

```diff
@@ -0,0 +1,271 @@
+interested in extracting data from the token.
+
+## Token Facts
+
+We want to encode certain information into the tokens.  The following attributes are defined:
+
+* `sentry_site`: references the target API URL that should be used.  A token will always have a
+  site in it and clients are not supposed to provide a fallback.  For instance this
+  would be `https://myorg.sentry.io/`.
+* `sentry_org`: a token is uniquely bound to an org, so the slug of that org is also always
+  contained.  Note that the slug is used rather than an org ID as the clients typically
+  need these slugs to create API requests.
+
+Potential fields:
+
+* `sentry_projects`: normally a token is valid for the entire org, but it could potentially
+  be restricted.  For operations such as source map uploads it might be interesting to issue
+  tokens bound to a single project in which case the upload experience does not require
+  providing the project slugs.  However we currently do not want to start with this.
+
+For JWT the facts could be encoded as custom claims:
+
+```json
+{
+    "iss": "sentry.io",
+    "iat": 1684154626,
+    "sentry_site": "https://myorg.sentry.io/",
+    "sentry_org": "myorg"
+}
+```
+
+Encoded the token then is be `sntrys_{encoded_jwt}`.
+
+## Transmitting Tokens
+
+Tokens are sent to the target sentry as `Bearer` token like normal.  The server uses the
+`sntrys_` prefix to automatically detect a structural token.  For existing tools that are
+unaware of the structure behind structural tokens nothing changes.
+
+## Parsing Tokens
+
+Clients are strongly encouraged to parse out the containing structure of the token and
+to use this information to route requests.  For the keys the following rules apply:
+
+* `sentry_site`: references the target API URL that should be used.  A token
+  will always have a site in it and clients are not supposed to provide an
+  automatic fallback.  If a site is not provided, one from the client config
+  should be picked up (typically `sentry.io`).
+* `org`: a token is uniquely bound to an org, so the slug of that org is also always
+  contained.  Note that the slug is used rather than an org ID as the clients typically
+  need these slugs to create API requests.
+
+An example of this with a JWT token:
+
+```python
+>>> import jwt
+>>> tok = "sntrys_eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzZW50cnkuaW8iLCJpYXQiOjE2ODQxNTQ2MjYsInNlbnRyeV9zaXRlIjoiaHR0cHM6Ly9teW9yZy5zZW50cnkuaW8vIiwic2VudHJ5X29yZyI6Im15b3JnIiwic2VudHJ5X3Byb2plY3RzIjpbIm15cHJvamVjdCJdfQ.ROnK3f72jGbH2CLkmswMIxXP1qZHDish9lN6kfCR0DU"
+>>> jwt.decode(tok[7:], options={"verify_signature": False})
+{
+  'iss': 'sentry.io',
+  'iat': 1684154626,
+  'sentry_site': 'https://myorg.sentry.io/',
+  'sentry_org': 'myorg',
+  'sentry_projects': ['myproject']  # not used currently
+}
+```
+
+## Token Issuance
+
+The purpose of this change is to allow any organization member to issue tokens with little
```

Comment:
I wonder if we should restrict _revocation_ to managers/owners? 🤔 this has the downside that a user cannot revoke a token they just created, but may be a reasonable safety net (even to disallow accidental deletion)?